### PR TITLE
Improve KnownBits interpreter & make it testable 

### PR DESCRIPTION
**Pull request by pranavk**
Work in progress ...

### CODE REVIEW

**Comment by regehr** [comment]
looks good, please squash

**Comment by regehr** [comment]
ah nevermind, I see it's a wip

**Comment by pranavk** [comment]
yeah, this PR is just for tracking for now; i will let you know when it's ready to merge.

File: include/souper/Infer/AbstractInterpreter.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#pragma once
```
Comment (regehr) [inline include/souper/Infer/AbstractInterpreter.h]: if we're going to do this, let's rather do it across the board in a separate PR

File: lib/Infer/AbstractInterpreter.cpp [hunk 2]
```diff
@@ -1,6 +1,272 @@
...
+    }
+
+    llvm::KnownBits mul(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(lhs.getBitWidth());
+
+      auto trailingZeros0 = lhs.countMinTrailingZeros();
+      auto trailingZeros1 = rhs.countMinTrailingZeros();
+      Result.Zero.setLowBits(std::min(trailingZeros0 + trailingZeros1, lhs.getBitWidth()));
+
+      // check for leading zeros
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: I think there are good opportunties for improvement here by counting leading ones as well as zeroes, right? no need to fix it now but please add a comment sketching out the necessary code

File: lib/Infer/AbstractInterpreter.cpp [hunk 3]
```diff
@@ -1,6 +1,272 @@
...
+      return Result;
+    }
+
+    llvm::KnownBits and_(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      auto Op0KB = lhs;
+      auto Op1KB = rhs;
+
+      Op0KB.One &= Op1KB.One;
+      Op0KB.Zero |= Op1KB.Zero;
+      return Op0KB;
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: isn't there a less awkward way to express this computation in a functional style, instead of creating these temporary objects and mutating them?

File: unittests/Interpreter/InterpreterTests.cpp [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+      return merge(bruteForce(x, setLowest(y), Pred),
+		   bruteForce(x, clearLowest(y), Pred));
+    auto xc = x.getConstant();
+    auto yc = y.getConstant();
+
+    KnownBits res(x.getBitWidth());
+    switch (Pred) {
+    case Inst::AddNUW:
+    case Inst::AddNW:
+    case Inst::Add:
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: where is addnsw?

File: lib/Infer/AbstractInterpreter.cpp (hunk 3 above)
Comment (pranavk) [reply lib/Infer/AbstractInterpreter.cpp]: i just copied them from the original file to here; will see what we can do here.

Comment (regehr) [reply lib/Infer/AbstractInterpreter.cpp]: it's fine for now if it's a copy

File: unittests/Interpreter/InterpreterTests.cpp (hunk 4 above)
Comment (pranavk) [reply unittests/Interpreter/InterpreterTests.cpp]: added now in latest commit

**Comment by regehr** [comment]
ok this is looking good

but please add a small unsoundness to one of the transfer functions and make sure that it is caught by the exhaustive testing framework, ok?

**Comment by regehr** [comment]
nothing about that will appear in a commit, of course-- it's just a sanity check you need to do before committing this

**Comment by pranavk** [comment]
> ok this is looking good
> 
> but please add a small unsoundness to one of the transfer functions and make sure that it is caught by the exhaustive testing framework, ok?

yup. i have done this in couple of transfer functions and the testing framework was able to catch the unsoundness. This is how the output looks like:

```
[pranavk@pranavk build]$ ./interpreter_tests 
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from InterpreterTests
[ RUN      ] InterpreterTests.KBTransferFunctions
Unsound!! mul
0001 mul 0001
Calculated: ???0
Expected: 0001
../unittests/Interpreter/InterpreterTests.cpp:376: Failure
Value of: testFn(Inst::Mul)
  Actual: false
Expected: true

```

**Comment by regehr** [comment]
perfect!

File: unittests/Interpreter/InterpreterTests.cpp [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+		       << Res2.getBitWidth() << ", Given: " << Res1.getBitWidth() << '\n';
+	  return false;
+	}
+	if (Res1.hasConflict() || Res2.hasConflict()) {
+	  llvm::errs() << "Expected or Given result has a conflict\n";
+	  return false;
+	}
+
+	for (unsigned i = 0; i < Res1.getBitWidth(); i++) {
+	  if ((Res1.Zero[i] and Res2.One[i]) || (Res1.One[i] and Res2.Zero[i])) {
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: let's avoid the alternative operators like "and" unless there's a compelling reason to use

File: unittests/Interpreter/InterpreterTests.cpp [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+	  Res1 = BinaryTransferFunctionsKB::slt(x, y);
+	  break;
+	case Inst::Ule:
+	  Res1 = BinaryTransferFunctionsKB::ule(x, y);
+	  break;
+	case Inst::Sle:
+	  Res1 = BinaryTransferFunctionsKB::sle(x, y);
+	  break;
+	default:
+	  break;
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: would be best to make this an llvm::report_fatal_error

Comment (regehr) [reply unittests/Interpreter/InterpreterTests.cpp]: try to avoid writing code that could hide a bug

File: unittests/Interpreter/InterpreterTests.cpp [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+    {
+      res = KnownBits(1);
+      if (xc.sle(yc))
+	res.One.setBit(0);
+      else
+	res.Zero.setBit(0);
+    }
+    break;
+    default:
+      break;
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: same-- this shouldn't get triggered, make it a fatal error

File: unittests/Interpreter/InterpreterTests.cpp [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+      if (xc.slt(yc))
+	res.One.setBit(0);
+      else
+	res.Zero.setBit(0);
+    }
+    break;
+    case Inst::Ule:
+    {
+      res = KnownBits(1);
+      if (xc.ule(yc))
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: can you just return the result of the ule call? why do you need to make a new object of the knownbits type?

File: unittests/Interpreter/InterpreterTests.cpp [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+      auto rc = xc | yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::Xor:
+    {
+      auto rc = xc ^ yc;
+      res.One = rc;
+      res.Zero = ~rc;
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: the repeated code for these is super ugly, why not just compute the APInt result for all operations and then at the bottom do the res.One and res.Zero thing?

File: unittests/Interpreter/InterpreterTests.cpp [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+    }
+    break;
+    case Inst::Mul:
+    {
+      auto rc = xc * yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+      break;
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: indentation

File: unittests/Interpreter/InterpreterTests.cpp [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+      auto rc = xc + yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::SubNUW:
+    case Inst::SubNSW:
+    case Inst::SubNW:
+    case Inst::Sub:
+    {
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: put this kind of curly brace on the line above after the :

File: unittests/Interpreter/InterpreterTests.cpp [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+  }
+
+  KnownBits setLowest(KnownBits x) {
+    for (int i = 0; i < x.getBitWidth(); i++) {
+      if (!x.Zero[i] && !x.One[i]) {
+	x.One.setBit(i);
+	return x;
+      }
+    }
+    assert(false);
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: this should be an llvm::report_fatal_error or llvm::unreachable

File: unittests/Interpreter/InterpreterTests.cpp [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+	x.Zero.clearBit(i);
+	x.One.setBit(i);
+	return true;
+      }
+      if (!x.Zero[i] && x.One[i]) {
+	x.Zero.clearBit(i);
+	x.One.clearBit(i);
+	continue;
+      }
+      assert(false && "error in nextKB");
```
Comment (regehr) [inline unittests/Interpreter/InterpreterTests.cpp]: report_fatal_error or unreachable

File: lib/Infer/AbstractInterpreter.cpp [hunk 14]
```diff
@@ -250,34 +438,38 @@ namespace souper {
...
-      }
+      return BinaryTransferFunctionsKB::ne(KB0, KB1);
+    }
+    case Inst::Ult: {
+      return BinaryTransferFunctionsKB::ult(KB0, KB1);
+    }
+    case Inst::Slt: {
+      return BinaryTransferFunctionsKB::slt(KB0, KB1);
+    }
+    case Inst::Ule: {
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: always avoid curly braces when not needed, like here

File: lib/Infer/AbstractInterpreter.cpp [hunk 15]
```diff
@@ -250,34 +438,38 @@ namespace souper {
       } else {
         return Result;
       }
-      if ((Other.Zero & 1) != 0) {
+
+      // Constants are never equal to 0
+      if (Other.Zero.isAllOnesValue()) {
         Result.Zero.setBit(0);
         return Result;
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: I know this isn't your code but please factor out the common "return Result"

File: lib/Infer/AbstractInterpreter.cpp [hunk 16]
```diff
@@ -1,6 +1,268 @@
...
+      llvm::KnownBits Result(1);
+      if (getSMax(lhs).slt(getSMin(rhs)))
+	Result.One.setBit(0);
+      if (getSMin(lhs).sge(getSMax(rhs)))
+	Result.Zero.setBit(0);
+      return Result;
+    }
+
+    llvm::KnownBits ule(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(1);
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: instead of creating a tempoarary object and mutating it, you should just construct the value that you want in the first place, there are several ways to do it, I think any is fine

Comment (regehr) [reply lib/Infer/AbstractInterpreter.cpp]: return APInt(1, 0) or return APInt(1, 1) should work just fine

File: lib/Infer/AbstractInterpreter.cpp [hunk 17]
```diff
@@ -1,6 +1,268 @@
...
+      unsigned Leaders =
+        std::max(lhs.countMinLeadingZeros(), rhs.countMinLeadingZeros());
+      Result.resetAll();
+      Result.Zero.setHighBits(Leaders);
+      return Result;
+    }
+
+    llvm::KnownBits and_(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      auto result = lhs;
+
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: don't put a blank line here if you don't have an analogous blank line in the next function

File: lib/Infer/AbstractInterpreter.cpp (hunk 16 above)
Comment (pranavk) [reply lib/Infer/AbstractInterpreter.cpp]: We are not returning APInt here; we are returning KnownBits. KnownBits doesn't have any public constructor that can let you do that; hence, the reason for this thing. There's a private constructor with which we can accomplish something like this but I think that would be much more unreadable than the current form.

File: unittests/Interpreter/InterpreterTests.cpp (hunk 8 above)
Comment (pranavk) [reply unittests/Interpreter/InterpreterTests.cpp]: result of the ule call is an APInt and KnownBits doens't have a public constructor that takes APInt.

File: lib/Infer/AbstractInterpreter.cpp (hunk 16 above)
Comment (regehr) [reply lib/Infer/AbstractInterpreter.cpp]: ugh.

**Comment by pranavk** [comment]
Okay. i have squashed it. feel free to merge.

**Comment by regehr** [comment]
LGTM!
